File: Algorithms_Java/Sort/QuickSort/QuickSort.java
/**Quick Sort
* @author keviness
* @version 2020/12/10
*/

import java.util.*;

public class QuickSort
{
    public static void main(String[] args)
    {
        double[] array = {-12, 3.4, 5.7, 9, 23, -0.3, 1.4};
        System.out.println("The source array:");
        show(array);
        System.out.println("The sorted array:");
        sort(array, 0, array.length-1);
        show(array);
    }

    public static void sort(double[] array, int low, int high)
    {
        int pos;
        if (low < high)
        {
            pos = findPos(array, low, high);
            sort(array, pos+1, high);    //the order can be changed!
            sort(array, low, pos-1);
        }
    }

    public static int findPos(double[] array, int low, int high)
    {
        double value = array[low];

        while (low < high)     //the order can't be changed!
        {
            while (low<high && value<=array[high])
            {
                high--;
            }
            array[low] = array[high];

            while (low<high && value>=array[low])
            {
                low++;
            }
            array[high] = array[low];
        }
        array[low] = value;

        return low;
    }

    public static void show(double[] array)
    {
        for (int i=0; i<array.length; i++)
        {
            System.out.printf("%4.2g ", array[i]);
        }
        System.out.printf("\n");
    }
}

File: Algorithms_C/Sorts/QuickSort.c
/*** quick sort ***/
#include <stdio.h>
#include <string.h>
#define SIZE 10
void quicksort(int *arr, int low, int high);
int findpos(int *arr, int low, int high);
void showarr(const int *arr);

int main(void)
{
    int arr[SIZE] = {-2, 1, 2, 3, 6, -7, 10, -9, -20, 20};
    showarr(arr);
    quicksort(arr, 0, 9);
    puts("\nThe arr content:");
    showarr(arr);

    return 0;
}

void quicksort(int *arr, int low, int high)
{
    int pos;
    if (low < high)
    {
        pos = findpos(arr, low, high);
        quicksort(arr, low, pos-1);
        quicksort(arr, pos+1, high);
    }
}

int findpos(int *arr, int low, int high)
{
    int value = arr[low];
    while (low < high)
    {
        while (low<high && arr[high]>=value)
        {
            high--;
        }
        arr[low] = arr[high];
        while (low<high && arr[low]<=value)
        {
            low++;
        }
        arr[high] = arr[low];
    }
    arr[low] = value;

    return low;
}

void showarr(const int *arr)
{
    for (int i=0; i<SIZE; i++)
    {
        printf("%d ", arr[i]);
    }
    putchar('\n');
}